### PR TITLE
MentionAvatars: Also display role icons in role mentions

### PR DESCRIPTION
**Pull request by SerStars**
Im still not sure about the name of the plugin but anyways

This plugin basically shows the role icon of the pinged roles (similar to [MentionAvatars](https://vencord.dev/plugins/MentionAvatars) by Ven)

Ported from BetterDiscord's [RoleMentionIcons](https://betterdiscord.app/plugin/RoleMentionIcons) by Neodymium

Screenshot:
Screenshot 2024-08-22 at 5 23 19 PM
"Retired Staff Member" has no role icon and shows the members icon in its role color
"R" has a role icon and it shows it

### CODE REVIEW

**Comment by SerStars** [comment]
heya 👋, 
i dont know if anyone saw this PR or not but ive seen PRs & issues still getting reviewed

**Comment by Scyye** [comment]
patience dear, patience

**Comment by SerStars** [comment]
> patience dear, patience

i know dear :>

**Comment by SerStars** [comment]
A SVG... here in PNG [Ven](https://discord.com/channels/1015060230222131221/1015063227299811479/1276623422289481813)
members

File: src/plugins/mentionRoleIcons/index.tsx [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    name: "MentionRoleIcons",
+    description: "Shows roles icons inside role mentions",
+    authors: [Devs.SerStars],
+
+    patches: [{
+        find: ".ROLE_MENTION)",
+        replacement: {
+            match: /function \i\(\i\){let{roleColor:.+?,roleId:(\i),.+?roleName:\i,guildId:(\i).+?}=\i,{analyticsLocations:.+\.ROLE_MENTION\),.+\i&&\(0,\i\.jsx\)\(\i\.RoleDot,{color:\(0,\i\.\i\).+?\.roleDot,background:.+?(1|0)}\),\i/,
+            replace: "$&,$self.AddIcon($1, $2)"
+        }
```
Comment (sadan4) [inline src/plugins/mentionRoleIcons/index.tsx:23]:
```suggestion
            find: ".ROLE_MENTION)",
            replacement: {
                match: /\.\.\.\i,children:\[.{0,100}\]/,
                replace: "$&.push($self.AddIcon(arguments[0]))"
            }
```
makes the patch a bit less hellish

File: src/plugins/mentionRoleIcons/index.tsx [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    }],
+    AddIcon(roleId, guildId) {
```
Comment (sadan4) [inline src/plugins/mentionRoleIcons/index.tsx:25]:
```suggestion
    AddIcon({ roleId, guildId }) {
```

**Review by sadan4 — COMMENTED** [review]
make the patch a bit less hellish

**Comment by SerStars** [comment]
Added the plugin to mentionAvatars as requested, it still works the same way (for both), but i couldnt figure out the patches to create an setting option to disable "@" before role mention
And i dont know if we are keeping the plugin name the same

**Comment by SerStars** [comment]
Issues should be resolved

File: src/plugins/mentionAvatars/index.tsx [hunk 3]
```diff
@@ -51,8 +58,21 @@ export default definePlugin({
                 {getUsernameString(username)}
             </span>
         );
-    }, { noop: true })
+    }, { noop: true }),
 
+    AddRoleIcon(roleId, guildId) {
+    const roleIcon = GuildStore.getRole(guildId, roleId);
```
Comment (Sqaaakoi) [inline src/plugins/mentionAvatars/index.tsx:64]: rename this to role

File: src/plugins/mentionAvatars/styles.css [hunk 4]
```diff
@@ -6,3 +6,11 @@
     border-radius: 50%;
     box-sizing: border-box;
 }
+
+.vc-mentionAvatars-roleicon {
+    vertical-align: middle;
+    width: 1em !important; /* just in case */
```
Comment (Sqaaakoi) [inline src/plugins/mentionAvatars/styles.css:12]: are you sure this !important is required? you already specify 16px in the attribute. pick one

File: src/plugins/mentionAvatars/index.tsx [hunk 5]
```diff
@@ -51,8 +58,21 @@ export default definePlugin({
...
+
+    if (roleIcon && roleIcon.icon) {
```
Comment (Sqaaakoi) [inline src/plugins/mentionAvatars/index.tsx:66]: why not do role?.icon

File: src/plugins/mentionAvatars/index.tsx [hunk 6]
```diff
@@ -51,8 +58,21 @@ export default definePlugin({
...
+        return (
+            <img className="vc-mentionAvatars-roleicon" src={`https://cdn.discordapp.com/role-icons/${roleId}/${roleIcon.icon}.webp?size=24&quality=lossless`} style={{ width: "16px", height: "16px" }} />
```
Comment (Sqaaakoi) [inline src/plugins/mentionAvatars/index.tsx:68]: style attribute is redundant?

File: src/plugins/mentionAvatars/index.tsx [hunk 7]
```diff
@@ -10,28 +10,35 @@ import { definePluginSettings } from "@api/Settings";
 import ErrorBoundary from "@components/ErrorBoundary";
 import { Devs } from "@utils/constants";
 import definePlugin, { OptionType } from "@utils/types";
-import { SelectedGuildStore, useState } from "@webpack/common";
+import { GuildStore,SelectedGuildStore, useState } from "@webpack/common";
```
Comment (Sqaaakoi) [inline src/plugins/mentionAvatars/index.tsx:13]: fix spacing here

File: src/plugins/mentionAvatars/README.md [hunk 8]
```diff
@@ -1,5 +1,6 @@
 # MentionAvatars
 
-Shows user avatars inside mentions
+Shows user avatars and role icons inside mentions
 
 ![](https://github.com/user-attachments/assets/fc76ea47-5e19-4063-a592-c57785a75cc7)
+![](https://github.com/user-attachments/assets/76c4c3d9-7cde-42db-ba84-903cbb40c163)
```
Comment (Sqaaakoi) [inline src/plugins/mentionAvatars/README.md:6]: do not replace the existing screenshot with a new screenshot that doesn't show the original functionality

File: src/plugins/mentionAvatars/index.tsx [hunk 9]
```diff
@@ -10,28 +10,35 @@ import { definePluginSettings } from "@api/Settings";
...
 import { User } from "discord-types/general";
 
 const settings = definePluginSettings({
     showAtSymbol: {
         type: OptionType.BOOLEAN,
-        description: "Whether the the @ symbol should be displayed",
+        description: "Whether the the @ symbol should be displayed on user mentions",
```
Comment (Sqaaakoi) [inline src/plugins/mentionAvatars/index.tsx:19]: create options for role mentions?

**Review by Sqaaakoi — COMMENTED** [review]
Is there a possibility to show the icon on the left like avatars, and replace the mention? Disclaimer: I haven't personally tested the code

**Comment by SerStars** [comment]
> and replace the mention?

What do you mean exactly with this?

File: src/plugins/mentionAvatars/index.tsx (hunk 9 above)
Comment (SerStars) [reply src/plugins/mentionAvatars/index.tsx:19]: That was the idea and will probably added i just couldn't figure out how to

File: src/plugins/mentionAvatars/styles.css (hunk 4 above)
Comment (SerStars) [reply src/plugins/mentionAvatars/styles.css:12]:
```suggestion
    width: 1em;
```
fine with this?

File: src/plugins/mentionAvatars/index.tsx (hunk 6 above)
Comment (SerStars) [reply src/plugins/mentionAvatars/index.tsx:68]: I'm sorry, I don't quite get what you mean

File: src/plugins/mentionAvatars/README.md (hunk 8 above)
Comment (SerStars) [reply src/plugins/mentionAvatars/README.md:6]: I didn't replace it, the original one is still there

File: src/plugins/mentionAvatars/index.tsx (hunk 9 above)
Comment (Vendicated) [reply src/plugins/mentionAvatars/index.tsx:19]: i think there's little reason to add it to role mentions considering their icons are at the back which doesn't "conflict" with the @

File: src/plugins/mentionAvatars/index.tsx [hunk 10]
```diff
@@ -10,28 +10,35 @@ import { definePluginSettings } from "@api/Settings";
...
         find: ".USER_MENTION)",
         replacement: {
             match: /children:"@"\.concat\((null!=\i\?\i:\i)\)(?<=\.useName\((\i)\).+?)/,
             replace: "children:$self.renderUsername({username:$1,user:$2})"
         }
+    },
+    {
+        find: ".ROLE_MENTION)",
+        replacement: {
+            match: /function \i\(\i\){let{roleColor:.+?,roleId:(\i),.+?roleName:\i,guildId:(\i).+?}=\i,{analyticsLocations:.+\.ROLE_MENTION\),.+\i&&\(0,\i\.jsx\)\(\i\.RoleDot,{color:\(0,\i\.\i\).+?\.roleDot,background:.+?(1|0)}\),\i/,
```
Comment (Vendicated) [inline src/plugins/mentionAvatars/index.tsx:39]: this seems very overcomplicated. match as little as possible. you can use `arguments` to get the argument without explicitly matching it

File: src/plugins/mentionAvatars/styles.css [hunk 11]
```diff
@@ -6,3 +6,11 @@
...
     box-sizing: border-box;
 }
+
+.vc-mentionAvatars-roleicon {
+    vertical-align: middle;
+    width: 1em;
+    height: 1em;
+    margin: 0 4px 0.2rem 2px;
+    box-sizing: border-box;
+}
```
Comment (Vendicated) [inline src/plugins/mentionAvatars/styles.css:16]: why not just reuse the other class? considering they're almost the same

File: src/plugins/mentionAvatars/index.tsx [hunk 12]
```diff
@@ -51,8 +58,47 @@ export default definePlugin({
                 {getUsernameString(username)}
             </span>
         );
-    }, { noop: true })
+    }, { noop: true }),
+
+    AddRoleIcon(roleId, guildId) {
```
Comment (Vendicated) [inline src/plugins/mentionAvatars/index.tsx:63]:
```suggestion
    renderRoleIcon: ErrorBoundary.wrap(({ roleId, guildId }) => {
```

**Comment by Vendicated** [comment]
please format your code better. keep lines a sane length and use newlines. you also seemingly haven't properly [configured your editor](https://docs.vencord.dev/installing/editor-setup/)

File: src/plugins/mentionAvatars/index.tsx (hunk 9 above)
Comment (SerStars) [reply src/plugins/mentionAvatars/index.tsx:19]: Is it even possible to move them to the left, ive tried so many times

**Comment by SerStars** [comment]
Should be resolved

**Pull request merged**